File: Data Structures & Algorithms/count-paths/submission-2.java
class Solution {

    public static int helper(int[][] dp,int m,int n,int x,int y){
        if(x==m-1 && y==n-1){
            return 1;
        }
        if(x>=m || y>=n){
            return 0;
        }
        if(dp[x][y]!=-1){
            return dp[x][y];
        }
        int w1=helper(dp,m,n,x,y+1);
        int w2=helper(dp,m,n,x+1,y);
        dp[x][y]=w1+w2;
        return w1+w2;
    }


    public int uniquePaths(int m, int n) {
        int[][] dp=new int[m][n];
        for(int[] i:dp){
            Arrays.fill(i,-1);
        }
        return helper(dp,m,n,0,0);
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)

        if n == 1:
            return nums[0]

        def solve(arr):
            m = len(arr)
            dp = [-1] * m
            def helper(curr):
                if curr >= m:
                    return 0
                if dp[curr] != -1:
                    return dp[curr]
                dp[curr] = arr[curr] + max(helper(curr + 2),helper(curr + 3))
                return dp[curr]
            return max(helper(0), helper(1))
        return max(solve(nums[:-1]),solve(nums[1:]))